src/rpc/xmlrpc_tinyxml2.cc

```diff
@@ -0,0 +1,451 @@
+        fault.as_map()["faultCode"] = XMLRPC_INTERNAL_ERROR;
+        result_list.push_back(fault);
+      }
+    }
+  } else {
+    result = execute_command(method_name, doc->FirstChildElement("methodCall")->FirstChildElement("params"));
+  }
+
+  printer->PushHeader(false, true);
+  printer->OpenElement("methodReponse", true);
+  printer->OpenElement("params", true);
+
+  printer->OpenElement("param", true);
+  printer->OpenElement("value", true);
+  print_object_xml(result, printer);
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+  tinyxml2::XMLDocument resultDoc;
+}
+
+
+void
+print_xmlrpc_fault(int faultCode, std::string faultString, tinyxml2::XMLPrinter* printer) {
+  printer->PushHeader(false, true);
+
+  printer->OpenElement("methodReponse", true);
+  printer->OpenElement("fault", true);
+  printer->OpenElement("struct", true);
+
+  printer->OpenElement("member", true);
+  printer->OpenElement("name", true);
+  printer->PushText("faultCode");
+  printer->CloseElement(true);
+  printer->OpenElement("value", true);
+  printer->OpenElement("int", true);
+  printer->PushText(faultCode);
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+
+  printer->OpenElement("member", true);
+  printer->OpenElement("name", true);
+  printer->PushText("faultString");
+  printer->CloseElement(true);
+  printer->OpenElement("value", true);
+  printer->OpenElement("string", true);
+  printer->PushText(faultString.c_str());
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+  printer->CloseElement(true);
+}
+
+bool
+XmlRpc::process(const char* inBuffer, uint32_t length, slot_write slotWrite) {
+  tinyxml2::XMLDocument doc;
+  doc.Parse(inBuffer, length);
+  try {
+    // This printer can't be reused in the 'catch' because while the
+    // buffer can be cleared, the internal stack of opened elements
+    // remains.
+    tinyxml2::XMLPrinter printer(nullptr, true, 0);
+    process_document(&doc, &printer);
+    return slotWrite(printer.CStr(), printer.CStrSize()-1);
+  } catch (xmlrpc_error& e) {
+    tinyxml2::XMLPrinter printer(nullptr, true, 0);
+    print_xmlrpc_fault(e.type(), e.what(), &printer);
+    return slotWrite(printer.CStr(), printer.CStrSize()-1);
+  } catch (torrent::local_error& e) {
+    tinyxml2::XMLPrinter printer(nullptr, true, 0);
+    print_xmlrpc_fault(XMLRPC_INTERNAL_ERROR, e.what(), &printer);
+    return slotWrite(printer.CStr(), printer.CStrSize()-1);
+  }
+}
+
+void XmlRpc::initialize() { m_isValid = true; }
+void XmlRpc::cleanup() {}
+
+void XmlRpc::insert_command(const char*, const char*, const char*) {}
+void XmlRpc::set_dialect(int) {}
+
+int64_t XmlRpc::size_limit() { return std::numeric_limits<int64_t>::max(); }
+void    XmlRpc::set_size_limit(uint64_t) {}
```

Comment:
Size limit needs to be a sane value, and in the same orders of magnitude that xmlrpc supports. (

Also, setting the size limit should have an effect, and be checked in `process(...)`.

Scgi sets `max_content_size    = (2 << 23);`.

Reply:
I genuinely can't imagine a situation where anyone would desire it to be anything other than the exact SCGI max length (no point is going higher, and content size is no predictor of CPU/memory usage, so going lower doesn't help even in low-spec situations), so I've had the default reference that value directly.

xmlrpc-c uses size_t for its limit, but I don't see a compelling reason to limit tinyxml2 to that, especially when it's capped by a hard-coded SCGI max content size in both cases anyway.